beacon_node/beacon_chain/src/attestation_verification/batch.rs

```diff
@@ -0,0 +1,229 @@
+//!
+//! In each function, attestations are "indexed" (i.e., the `IndexedAttestation` is computed), to
+//! determine if they should progress to signature verification. Then, all attestations which were
+//! successfully indexed have their signatures verified in a batch. If that signature batch fails
+//! then all attestation signatures are verified independently.
+//!
+//! The outcome of each function is a `Vec<Result>` with a one-to-one mapping to the attestations
+//! supplied as input. Each result provides the exact success or failure result of the corresponding
+//! attestation, with no loss of fidelity when compared to individual verification.
+use super::{
+    CheckAttestationSignature, Error, IndexedAggregatedAttestation, IndexedUnaggregatedAttestation,
+    VerifiedAggregatedAttestation, VerifiedUnaggregatedAttestation,
+};
+use crate::{
+    beacon_chain::{HEAD_LOCK_TIMEOUT, VALIDATOR_PUBKEY_CACHE_LOCK_TIMEOUT},
+    metrics, BeaconChain, BeaconChainError, BeaconChainTypes,
+};
+use bls::verify_signature_sets;
+use state_processing::signature_sets::{
+    indexed_attestation_signature_set_from_pubkeys, signed_aggregate_selection_proof_signature_set,
+    signed_aggregate_signature_set,
+};
+use std::borrow::Cow;
+use types::*;
+
+/// Verify aggregated attestations using batch BLS signature verification.
+///
+/// See module-level docs for more info.
+pub fn batch_verify_aggregated_attestations<'a, T, I>(
+    aggregates: I,
+    chain: &BeaconChain<T>,
+) -> Result<Vec<Result<VerifiedAggregatedAttestation<'a, T>, Error>>, Error>
+where
+    T: BeaconChainTypes,
+    I: Iterator<Item = &'a SignedAggregateAndProof<T::EthSpec>> + ExactSizeIterator,
+{
+    let mut num_indexed = 0;
+    let mut num_failed = 0;
+
+    // Perform indexing of all attestations, collecting the results.
+    let indexing_results = aggregates
+        .map(|aggregate| {
+            let result = IndexedAggregatedAttestation::verify(aggregate, chain);
+            if result.is_ok() {
+                num_indexed += 1;
+            } else {
+                num_failed += 1;
+            }
+            result
+        })
+        .collect::<Vec<_>>();
+
+    // May be set to `No` if batch verification succeeds.
+    let mut check_signatures = CheckAttestationSignature::Yes;
+
+    // Perform batch BLS verification, if any attestation signatures are worth checking.
+    if num_indexed > 0 {
+        let signature_setup_timer =
+            metrics::start_timer(&metrics::ATTESTATION_PROCESSING_SIGNATURE_SETUP_TIMES);
+
+        let pubkey_cache = chain
+            .validator_pubkey_cache
+            .try_read_for(VALIDATOR_PUBKEY_CACHE_LOCK_TIMEOUT)
+            .ok_or(BeaconChainError::ValidatorPubkeyCacheLockTimeout)?;
+
+        let fork = chain
+            .canonical_head
+            .try_read_for(HEAD_LOCK_TIMEOUT)
+            .ok_or(BeaconChainError::CanonicalHeadLockTimeout)
+            .map(|head| head.beacon_state.fork())?;
```

Comment:
```suggestion
            let fork = chain.with_head(|head| Ok::<_, BeaconChainError>(head.beacon_state.fork()))?;
```

I think it makes sense to avoid referring to the head lock timeouts explicitly, in case we change how that works behind the scenes